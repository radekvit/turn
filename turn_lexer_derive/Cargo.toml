[package]
name = "turn_lexer_derive"
version = "0.0.0"
license = "MIT"
authors = ["Radek Vit <radekvitr@gmail.com>"]
description = "Implementation of #[derive(Lexer)]"
keywords = ["turn", "lexer", "lexical analyzer"]
readme = "crates-io.md"
edition = "2018"

[lib]
name = "turn_lexer_derive"
proc-macro = true

[[test]]
name = "tests"
path = "test/tests.rs"

[dev-dependencies]
trybuild = "1.0"

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
syn = "1.0"
