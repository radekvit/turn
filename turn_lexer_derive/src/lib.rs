#![allow(dead_code)]

extern crate proc_macro;

mod automata;
mod derive_parse;
mod matchers;
//mod lexer_impl;

use proc_macro::TokenStream;
use syn::{parse_macro_input, DeriveInput};
//use proc_macro2::TokenStream as TokenStream2;
//use quote::{quote, quote_spanned};
//use syn::{parse_macro_input, DeriveInput, Ident, Data, Fields, Type,PathArguments, GenericArgument,
//    Attribute, Meta, Lit};

#[proc_macro_derive(Lexer, attributes(skip, token, regex))]
pub fn derive(input: TokenStream) -> TokenStream {
    // parse the derive input and process all attributes
    let input = match derive_parse::parse(parse_macro_input!(input as DeriveInput)) {
        Ok(input) => input,
        Err(error) => return error.to_compile_error().into(),
    };
    // create a minimal finite state automaton from the input
    let _automaton = match automata::create_minimal_automaton(&input) {
        Ok(automaton) => automaton,
        Err(error) => return error.to_compile_error().into(),
    };
    // create turn::Lexer implementation for this enum
    //lexer_impl::create_implementation(automaton)
    TokenStream::new()
}
