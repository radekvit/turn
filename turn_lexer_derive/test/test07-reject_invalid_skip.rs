use turn_lexer_derive::Lexer;

#[derive(Lexer)]
#[skip = 420]
enum Foo {
    #[token = "one"]
    One,
    #[regex = "[tT]wo."]
    Two,
}

#[derive(Lexer)]
#[skip ("420", "b")]
enum Bar {
    #[token = "one"]
    One,
    #[regex = "[tT]wo."]
    Two,
}

#[derive(Lexer)]
#[skip]
enum Baz {
    #[token = "one"]
    One,
    #[regex = "[tT]wo."]
    Two,
}

fn main() {}
