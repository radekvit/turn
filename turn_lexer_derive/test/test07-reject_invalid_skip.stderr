error: Attribute skip must be a string or char literal.
 --> $DIR/test07-reject_invalid_skip.rs:4:3
  |
4 | #[skip = 420]
  |   ^^^^

error: The skip regex must be defined as #[skip = "..."]
  --> $DIR/test07-reject_invalid_skip.rs:13:3
   |
13 | #[skip ("420", "b")]
   |   ^^^^

error: The skip regex must be defined as #[skip = "..."]
  --> $DIR/test07-reject_invalid_skip.rs:22:3
   |
22 | #[skip]
   |   ^^^^
